### PR TITLE
set controller as service

### PR DESCRIPTION
**Pull request by oallain**
Feature for "Define controllers as services" #91

### CODE REVIEW

**Comment by oallain** [comment]
Hello @stefandoorn 

Does this PR match your expectation for issue #91?

**Comment by lsmith77** [comment]
thank you for working on this!

I would want to try to avoid to inject the container, if at all we should inject a service locator. however I think it might be even better if we inject a service that has the responsibility to determine the entity name and that service uses a service locator.

https://symfony.com/doc/current/service_container/service_subscribers_locators.html#defining-a-service-locator

**Comment by oallain** [comment]
Hello @lsmith77,

I would like too, but the ParameterBag only became a service from Symfony 4.1 (see 
https://symfony.com/blog/new-in-symfony-4-1-getting-container-parameters-as-a-service), so i haven't any other simple solution to get parameter dynamically.

Do you have an idea to get workaround this?

**Comment by lsmith77** [comment]
@oallain the workaround I am proposing is to define a service that implements `ParameterBagInterface` of our own that does have the container injected to fetch parameters. this way we have at least separated this "uglyness" from the controller.

since `Symfony\Component\DependencyInjection\ParameterBag\ParameterBagInterface` has been around a long time, we can easily type hint it and then have some BC logic inside our DI FOSSyliusImportExportExtension to determine if we need to inject our BC service or can rely on Symfony autowiring.

does that make sense?

**Comment by oallain** [comment]
@lsmith77 job done :+1:

File: src/Resources/config/services.yml [hunk 1]
```diff
@@ -1,13 +1,20 @@
 services:
+    # Parameter Bag
```
Comment (lsmith77) [inline src/Resources/config/services.yml]: @solutionDrive-Alt / @stefandoorn wdyt should we try to use the core feature for 4.1 https://symfony.com/blog/new-in-symfony-4-1-getting-container-parameters-as-a-service ?

this would require some extra work in the DI Extension to determine the Symfony version to decide if autowiring will already inject the parameter bag automatically or not.

File: src/Controller/ExportDataController.php [hunk 2]
```diff
@@ -84,15 +101,14 @@ private function exportData(Request $request, string $exporter, string $format,
...
 
-        /** @var \Sylius\Component\Resource\Repository\RepositoryInterface $repository */
-        $repository = $this->get($repositoryName);
+        if (!$this->parameterBag->has($parameter)) {
+            throw new \Exception(sprintf("Parameter '%s' is not defined", $parameter));
+        }
+        $entityName = $this->parameterBag->get($parameter);
 
-        return $repository;
+        return new EntityRepository($this->entityManager, new ClassMetadata($entityName));
```
Comment (lsmith77) [inline src/Controller/ExportDataController.php]: I think we should rather use `$this->entityManager->getRepository($entityName);` here

see https://github.com/doctrine/persistence/blob/master/lib/Doctrine/Persistence/ObjectManager.php#L110

that being said previously we fetched the repository as a service which of course is also an extension point, which we now side-step. I wonder if Sylius provides a factory service that receives the `$entityName` and returns the given `EntityRepository` instance?

Comment (lsmith77) [reply src/Controller/ExportDataController.php]: posted a question about this in #general on Sylius slack

File: src/Resources/config/services.yml (hunk 1 above)
Comment (solutionDrive-Alt) [reply src/Resources/config/services.yml]: Sounds interesting. But probably something for a different PR.

File: src/Controller/ExportDataController.php (hunk 2 above)
Comment (lsmith77) [reply src/Controller/ExportDataController.php]: maybe using `$em->getRepository()` can be considered safe .. confirming here https://github.com/Sylius/Sylius/issues/873

Comment (lsmith77) [reply src/Controller/ExportDataController.php]: @oallain if you have time .. maybe you can test if one really gets the same instance via `$em->getRepository()` as via `$container->get()` .. because if we do, then imho we can wrap this PR up.

File: src/DependencyInjection/ParameterBag/ParameterBag.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<?php
```
Comment (lsmith77) [inline src/DependencyInjection/ParameterBag/ParameterBag.php]: maybe add a comment here that this is a BC solution only needed while we support versions below 4.1 and add a link to https://symfony.com/blog/new-in-symfony-4-1-getting-container-parameters-as-a-service

File: src/Controller/ExportDataController.php (hunk 2 above)
Comment (oallain) [reply src/Controller/ExportDataController.php]: Hi @lsmith77 ,

Work :

- `$this->entityManager->getRepository(\Sylius\Component\Core\Model\Order::class)`

Don't work :

- `$this->entityManager->getRepository('Order')`
- `$this->entityManager->getRepository('sylius.model.order.class')`

But, I just have the class name like "country", "order" ....

So, and don't know how to make it better :(

File: src/DependencyInjection/ParameterBag/ParameterBag.php (hunk 3 above)
Comment (oallain) [reply src/DependencyInjection/ParameterBag/ParameterBag.php]: :+1:

File: src/Controller/ExportDataController.php (hunk 2 above)
Comment (lsmith77) [reply src/Controller/ExportDataController.php]:
```
pjedrzejewski   [24 minutes ago]
I don’t think there is other way now… except making a “repository locator” which would be a container too but just smaller


pjedrzejewski   [24 minutes ago]
the genericness of this controller requires a container of some sort imo

pjedrzejewski   [23 minutes ago]
that’s why in the ResourceBundle we generate controller per resource, where we inject the right repository

lsmith   [20 minutes ago]
ok .. I guess we could do the same .. would of course increase number of routes and services .. but I guess the overhead is small .. and as an upside it adds an extension point by resource
```

@stefandoorn / @solutionDrive-Alt wdyt? should we provide a controller per resource?

Comment (lsmith77) [reply src/Controller/ExportDataController.php]: https://sylius-devs.slack.com/archives/C3EGDG9LY/p1552896248352900

Comment (solutionDrive-Alt) [reply src/Controller/ExportDataController.php]: I think when the importers / exporters get more sophisticated, we won't be able to handle it in that generic way anyways. On the other hand we will have to be carefull, that it don't evolves into a mess, where every Export/Import-Controller does the same thing differently.

Comment (oallain) [reply src/Controller/ExportDataController.php]: Hello @lsmith77 and @solutionDrive-Alt ,

What do you want to do?
We validate like this, or do we do otherwise?
 
Thanks

Comment (lsmith77) [reply src/Controller/ExportDataController.php]: so imho we should create separate services for each route allowing us to inject the specific repository instance rather than having to fetch it inside the controller.

so @oallain if you are motivated, we would appreciate it if you could do the necessary adjustments. thank for all your work!

Comment (oallain) [reply src/Controller/ExportDataController.php]: Ok, i'll try to work on it this week.

File: src/Controller/ExportDataController.php [hunk 4]
```diff
@@ -35,11 +44,17 @@ final class ExportDataController extends Controller
     public function __construct(
         ServiceRegistryInterface $registry,
         RequestConfigurationFactoryInterface $requestConfigurationFactory,
-        ResourcesCollectionProviderInterface $resourcesCollectionProvider
+        ResourcesCollectionProviderInterface $resourcesCollectionProvider,
+        RepositoryInterface $repository,
+        EntityManager $entityManager,
```
[the remark this answers is not in the document]
Comment (oallain) [reply src/Controller/ExportDataController.php]: Yes, sorry

File: src/Resources/config/services.yml [hunk 5]
```diff
@@ -1,21 +1,56 @@
...
+
+    sylius.controller.export_data_country:
+        public: true
+        class: FriendsOfSylius\SyliusImportExportPlugin\Controller\ExportDataController
+        arguments:
+            - "@sylius.exporters_registry"
+            - "@sylius.resource_controller.request_configuration_factory"
+            - "@sylius.resource_controller.resources_collection_provider"
+            - "@sylius.repository.country"
+            - "@doctrine.orm.entity_manager"
```
Comment (lsmith77) [inline src/Resources/config/services.yml]: entity manager also needs to be removed from all the relevant services in this file

Comment (oallain) [reply src/Resources/config/services.yml]: Yes, i was sleeping when Travis get result ;)

Comment (oallain) [reply src/Resources/config/services.yml]: Fixed :)

Comment (lsmith77) [reply src/Resources/config/services.yml]: and green we are!

Comment (oallain) [reply src/Resources/config/services.yml]: Yeah !

**Comment by lsmith77** [comment]
one last thing would be adding something to the README.md to remind users to add those controller services as needed. while I realized that I also remembered that we could make use of service inheritance to make it easier to define those resource specific services https://symfony.com/doc/current/service_container/parent_services.html

that being said, we could just as well make this a follow up ticket.

**Comment by lsmith77** [comment]
@stefandoorn / @solutionDrive-Alt can you review / merge ?

As noted above, from the code it looks good to me .. but I didn't run any manual tests to confirm if things are working.

File: README.md [hunk 6]
```diff
@@ -308,11 +308,26 @@ app_export_data_foo:
     methods: [GET]
     defaults:
         resource: sylius.foo
-        _controller: sylius.controller.export_data:exportAction
+        _controller: sylius.controller.export_data_foo:exportAction
         _sylius:
             filterable: true
             grid: sylius_admin_foo # Name of defined grid here
 ```
+and the associate controller
```
Comment (stefandoorn) [inline README.md]: I would change in something like: `And add the associated controller service definition to your services`.

File: README.md [hunk 7]
```diff
@@ -368,12 +383,26 @@ app_export_data_country:
...
         resource: sylius.country
-        _controller: sylius.controller.export_data:exportAction
+        _controller: sylius.controller.export_data_contry:exportAction
         _sylius:
             filterable: true
             grid: sylius_admin_country
 ```   
-   
+
+and the associate controller
```
Comment (stefandoorn) [inline README.md]: Same here

File: src/Controller/ExportDataController.php [hunk 8]
```diff
@@ -16,13 +16,18 @@
...
-use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\HttpFoundation\ResponseHeaderBag;
 
-final class ExportDataController extends Controller
+final class ExportDataController
 {
+    /** @var array */
+    private $syliusResources;
```
Comment (stefandoorn) [inline src/Controller/ExportDataController.php]: What about renaming to `$resources`? We are in the context of Sylius.

File: src/DependencyInjection/ParameterBag/ParameterBag.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+declare(strict_types=1);
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\DependencyInjection\ParameterBag;
+
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\DependencyInjection\ParameterBag\FrozenParameterBag;
+use Symfony\Component\DependencyInjection\ParameterBag\ParameterBagInterface;
+
+class ParameterBag extends FrozenParameterBag implements ParameterBagInterface
```
Comment (stefandoorn) [inline src/DependencyInjection/ParameterBag/ParameterBag.php]: Shall we make this one final?

File: src/DependencyInjection/ParameterBag/ParameterBag.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is need while SyliusImportExportPlugin support Symfony version < 4.1
```
Comment (stefandoorn) [inline src/DependencyInjection/ParameterBag/ParameterBag.php]: I would reword a bit here:

```
This is file is required as long as SyliusImportExportPlugin supports Symfony versions lower than 4.1. Before this version, ParameterBag wasn't a service available.

See: https://symfony.com/blog/new-in-symfony-4-1-getting-container-parameters-as-a-service
```

File: src/Resources/config/routing.yml [hunk 11]
```diff
@@ -7,31 +7,31 @@ app_import_data:
 
 ## Export routes for specific settings
 app_export_data_country:
-    path: /export/sylius.country/{format}
+    path: /export/{resource}/{format}
```
Comment (stefandoorn) [inline src/Resources/config/routing.yml]: Is it handy to have `{resource}` here, while in the `_controller` we 'nail' it to the `export_data_country` controller? I think then we should be explicit in both.

File: src/Resources/config/routing.yml [hunk 12]
```diff
@@ -7,31 +7,31 @@ app_import_data:
...
         resource: sylius.country
-        _controller: sylius.controller.export_data:exportAction
+        _controller: sylius.controller.export_data_country:exportAction
         _sylius:
             filterable: true
             grid: sylius_admin_country
 
 app_export_data_order:
-    path: /export/sylius.order/{format}
+    path: /export/{resource}/{format}
```
Comment (stefandoorn) [inline src/Resources/config/routing.yml]: Same here

File: src/Resources/config/routing.yml [hunk 13]
```diff
@@ -7,31 +7,31 @@ app_import_data:
...
         resource: sylius.order
-        _controller: sylius.controller.export_data:exportAction
+        _controller: sylius.controller.export_data_order:exportAction
         _sylius:
             filterable: true
             grid: sylius_admin_order
 
 app_export_data_customer:
-    path: /export/sylius.customer/{format}
+    path: /export/{resource}/{format}
```
Comment (stefandoorn) [inline src/Resources/config/routing.yml]: Same here

File: README.md (hunk 6 above)
Comment (oallain) [reply README.md]: fixed

File: src/Resources/config/routing.yml (hunk 11 above)
Comment (oallain) [reply src/Resources/config/routing.yml]: Yes, it's more explicit ;)

Comment (stefandoorn) [reply src/Resources/config/routing.yml]: @lsmith77 @solutionDrive-Alt WDYT?

Comment (stefandoorn) [reply src/Resources/config/routing.yml]: Basically what I meant is, shouldn't we just keep `sylius.country` there? It's not variable, as it's tied to the country controller specific.

Comment (lsmith77) [reply src/Resources/config/routing.yml]: while the service specific, the controller code is generic. so in that sense its nice to have easy access to the "context" of which resource is being handled by this specific instance of the generic controller. that being said, when customizing the controller one has different ways to get this info:

1) by introspecting the repository instance
2) by adding another service parameter

I am unsure if these approaches are less/more elegant.
One use case that cannot be covered by them would be event listeners that get triggered before the controller is called. If those are legitimate I am not sure.

So to conclude, we could remove this placeholder for now simply because we do not yet have a concrete use case (unless someone has one). That being said, leaving it as is, imho is also not very risky. As such I am fine either or.

Comment (stefandoorn) [reply src/Resources/config/routing.yml]: Then I propose to leave it as in the PR for now, as I think later on we might be able to do something neat with registration based on tags instead of each controller separate.

**Comment by lsmith77** [comment]
wow thanks so much @oallain !

**Comment by oallain** [comment]
Thanks too ! :tada:
